src/wh_client_crypto.c

```diff
@@ -156,6 +157,284 @@ int wh_Client_RngGenerate(whClientContext* ctx, uint8_t* out, uint32_t size)
     return ret;
 }
 
+#ifndef NO_AES
+
+#ifdef HAVE_AES_CBC
+int wh_Client_AesCbc(whClientContext* ctx,
+        Aes* aes, int enc,
+        uint32_t len, const uint8_t* in,
+        uint8_t* out)
+{
+    int ret = 0;
+    uint16_t group      = WH_MESSAGE_GROUP_CRYPTO;
+    uint16_t action     = WC_ALGO_TYPE_CIPHER;
+    uint16_t type       = WC_CIPHER_AES_CBC;
+    uint16_t dataSz     = 0;
+
+    uint16_t blocks = len / AES_BLOCK_SIZE;
+    uint32_t last_offset = (blocks - 1) * AES_BLOCK_SIZE;
+
+
+    uint32_t key_len    = aes->keylen;
+    const uint8_t* key  = (const uint8_t*)(aes->devKey);
+    whKeyId key_id      = WH_DEVCTX_TO_KEYID(aes->devCtx);
+
+    whPacket* packet    = (whPacket*)wh_CommClient_GetDataPtr(ctx->comm);
+
+    uint8_t* iv         = (uint8_t*)aes->reg;
+    uint32_t iv_len     = AES_IV_SIZE;
+
+    /* Request packet */
+    wh_Packet_cipher_aescbc_req* req = &packet->cipherAesCbcReq;
+    uint8_t* req_in     = (uint8_t*)(req + 1);
+    uint8_t* req_key    = req_in + len;
+    uint8_t* req_iv     = req_key + key_len;
+
+    uint32_t req_len    = WH_PACKET_STUB_SIZE + sizeof(*req) + len +
+                            key_len + iv_len;
+
+    /* Response packet */
+    wh_Packet_cipher_aescbc_res* res = &packet->cipherAesCbcRes;
+    uint8_t* res_out    = (uint8_t*)(res + 1);
+
+    uint32_t res_len    = WH_PACKET_STUB_SIZE+ sizeof(*res) + len;
+
+#ifdef DEBUG_CRYPTOCB_VERBOSE
+    printf("[client] %s: enc:%d keylen:%d ivsz:%d insz:%d reqsz:%u ressz:%u "
+                      "blocks:%u lastoffset:%u\n",
+                __func__,enc, key_len, iv_len, len,  req_len, res_len,
+                blocks, last_offset);
+    printf("[client] %s: req:%p in:%p key:%p iv:%p res:%p out:%p\n",
+            __func__, req, req_in, req_key, req_iv, res, res_out);
+#endif
+
+    /* III 0 size check is done in wolfCrypt */
```

Comment:
Just a thought addressing the TODO - I'd say keep the check in, I'd rather do redundant checking of args with a microscopic perf penalty, but keep functions safe for use outside the cryptoCallback.

Reply:
Updated to handle locally.